brayns/engine/scenecomponents/SceneModelManager.h

```diff
@@ -0,0 +1,177 @@
+     */
+    ModelInstance &getModelInstance(const uint32_t instanceID);
+
+    /**
+     * @brief Return a list of all model instances in the manager
+     * @return std::vector<ModelInstance *> &
+     */
+    std::vector<ModelInstance *> &getAllModelInstances() noexcept;
+
+    /**
+     * @brief Return a list of all model instances in the manager
+     * @returns const std::vector<ModelInstance *> &
+     */
+    const std::vector<ModelInstance *> &getAllModelInstances() const noexcept;
+
+    /**
+     * @brief Removes a model instance from the scene, identified by the given instance ID.
+     * If the model to which the instance refers does not have any other instance, it will be deleted as well.
+     * @param instanceID the ID of the instance to re,pve
+     * @throws std::invalid_argument if modelID does not correspond to any existing model
+     */
+    void removeModel(const uint32_t instanceID);
+
+    /**
+     * @brief Returns the parameters with which a model referenced by the given instance was loaded
+     * @param instanceID ID of the instance that owns the model to fetch the load parameters
+     * @return ModelParameters &
+     * @throws std::invalid_argument if the given instanceID does not belong to any existing instance
+     */
+    const ModelLoadParameters &getModelLoadParameters(const uint32_t instanceID) const;
+
+private:
+    friend class Scene;
+
+    /**
+     * @brief Calls the preRender function on all components of all models
+     *
+     * @param parameters
+     */
+    void preRender(const ParametersManager &parameters);
+
+    /**
+     * @brief Attempts to commit any unsynced data to OSPRay.
+     *
+     * @return true if anything was commited to OSPRay, false otherwise
+     */
+    bool commit();
+
+    /**
+     * @brief Calls the postRender function on all componnets of all models
+     *
+     * @param parameters
+     */
+    void postRender(const ParametersManager &parameters);
+
+    /**
+     * @brief Queries the bounds of all model instances and returns a Bounds object encapsulating all of them
+     *
+     * @return Bounds
+     */
+    Bounds getBounds() const noexcept;
+
+    /**
+     * @brief Return a list with all the model instance handles so that the scene can commit them
+     *
+     * @return std::vector<OSPInstance>
+     */
+    std::vector<OSPInstance> getInstanceHandles() noexcept;
+
+    /**
+     * @brief returns the size in bytes of this model manager
+     * @return size_t
+     */
+    size_t getSizeInBytes() const noexcept;
+
+private:
+    /**
+     * @brief The ModelEntry struct holds all the information related to a model
+     * It allows easy access to the models of the scene without having to account for instancing
+     */
+    struct ModelEntry
+    {
+        ModelLoadParameters params;
+        std::unique_ptr<Model> model;
+        std::vector<std::unique_ptr<ModelInstance>> instances;
```

Comment:
Maybe by value ?

Reply:
Cannot be replaced by value: ModelInstance has a reference member, thus implicitly deleting default, copy and move constructors.